apis/v1alpha1/tlsconfig_types.go

```diff
@@ -65,6 +65,18 @@ type TLSConfig struct {
 	//
 	// +required
 	CertificateRef CertificateObjectReference `json:"certificateRef,omitempty"`
+
+	// AllowOverrideViaRoute dictates if TLS certificates can be configured
+	// via Routes or not. If set to true, a TLS certificate for a hostname
+	// defined in  Route takes precedence over the certificate defined in
+	// Gateway.
+	// CertificateRef must be defined even if this is set to true as it will
+	// be used as the default certificate for the listener.
+	//
+	// +kubebuilder:default=false
+	// +optional
+	AllowOverrideViaRoute bool `json:"allowOverrideViaRoute,omitempty"`
```

Comment:
Not sure if this is actually better, but would `AllowRouteOverride` work here?

Reply:
Sure, that sounds pleasant to my ears as well. Will update shortly.

Reply:
What if this field was changed from a bool to a struct that provides additional override flexibility? Something like:

```
// RouteOverrideType defines a schema for overriding gateway certificates.
// +kubebuilder:validation:Enum=AllowAll,DenyAll,AllowNamespaces
// +kubebuilder:default=DenyAll
type RouteOverrideType string

const (
	// AllowAll allows cert overrides for all routes.
	AllowAll RouteOverrideType = "AllowAll"

	// Deny denies cert overrides for all routes.
	DenyAll RouteOverrideType = "DenyAll"

	// AllowNamespaces allows cert overrides for routes in specified namespaces.
	AllowNamespaces RouteOverrideType = "AllowNamespaces"
)

// CertOverridePolicy defines a schema for overriding gateway certificates.
type CertOverridePolicy struct {
	// Type is the type of cert override policy for this gateway to use.
	//
	// Support: Core
	//
	// +optional
	// +kubebuilder:default=DenyAll
	Type RouteOverrideType `json:"type"`

        // namespaces is the list of namespaces where routes can override gateway certs.
	Namespaces []string `json:"namespaces"`

type TLSConfig struct {
...
CertOverridePolicy CertOverridePolicy `json:"certOverridePolicy"`
...
}

Reply:
We can make this more flexible if there is a need for this although we are going to sacrifice on simplicity quite a bit.
Since TLS is a sensitive piece of configuration, we should aim to keep things a little dumb, where possible, to minimize the number of bugs in implementations and end-user interpretation of the API.

I see the reasoning for adding what you are proposing but I'm concerned about the complexity here.
I'm totally open to being convinced otherwise here!

Reply:
I've updated the PR with some parts of what @danehans has proposed. Please take a look.